applications/ContactStructuralMechanicsApplication/custom_utilities/contact_utilities.cpp

```diff
@@ -0,0 +1,322 @@
+// KRATOS  ___|  |                   |                   |
+//       \___ \  __|  __| |   |  __| __| |   |  __| _` | |
+//             | |   |    |   | (    |   |   | |   (   | |
+//       _____/ \__|_|   \__,_|\___|\__|\__,_|_|  \__,_|_| MECHANICS
+//
+//  License:             BSD License
+//                                       license: StructuralMechanicsApplication/license.txt
+//
+//  Main authors:    Vicente Mataix Ferrandiz
+//
+
+// System includes
+
+// External includes
+
+// Project includes
+#include "utilities/openmp_utils.h"
+#include "utilities/math_utils.h"
+#include "includes/model_part.h"
+#include "custom_utilities/contact_utilities.h"
+
+namespace Kratos
+{
+double ContactUtilities::CalculateRelativeSizeMesh(ModelPart& rModelPart)
+{
+    return CalculateMaxNodalH(rModelPart)/CalculateMinimalNodalH(rModelPart);
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+double ContactUtilities::CalculateMaxNodalH(ModelPart& rModelPart)
+{
+    // We iterate over the nodes
+    NodesArrayType& r_nodes_array = rModelPart.Nodes();
+    const auto it_node_begin = r_nodes_array.begin();
+
+//     // Creating the max auxiliar value
+//     double max_value = 0.0;
+//
+//     #pragma omp parallel for reduction(max:max_value)
+//     for(int i = 0; i < static_cast<int>(r_nodes_array.size()); ++i) {
+//         auto it_node = it_node_begin + i;
+//         KRATOS_DEBUG_ERROR_IF_NOT(it_node->SolutionStepsDataHas(NODAL_H)) << "ERROR:: NODAL_H not added" << std::endl;
+//         max_value = std::max(max_value, it_node->FastGetSolutionStepValue(NODAL_H));
+//     }
+//
+//     return max_value;
+
+    // Creating a buffer for parallel vector fill
+    const int num_threads = OpenMPUtils::GetNumThreads();
+    std::vector<double> max_vector(num_threads, 0.0);
+    double nodal_h;
+    #pragma omp parallel for private(nodal_h)
+    for(int i = 0; i < static_cast<int>(r_nodes_array.size()); ++i) {
+        auto it_node = it_node_begin + i;
+        KRATOS_DEBUG_ERROR_IF_NOT(it_node->SolutionStepsDataHas(NODAL_H)) << "ERROR:: NODAL_H not added" << std::endl;
+        nodal_h = it_node->FastGetSolutionStepValue(NODAL_H);
+
+        const int id = OpenMPUtils::ThisThread();
+
+        if (nodal_h > max_vector[id])
+            max_vector[id] = nodal_h;
+    }
+
+    return *std::max_element(max_vector.begin(), max_vector.end());
```

Comment:
just to say that this could be done simpler with the ParallelUtilities

Reply:
I have it in my TODO list